firedrake/adjoint/dirichletbc.py

```diff
@@ -20,14 +20,17 @@ def wrapper(self, *args, **kwargs):
 
     @staticmethod
     def _ad_annotate_apply(apply):
-        @no_annotations
         @wraps(apply)
         def wrapper(self, *args, **kwargs):
-            for arg in args:
-                if not hasattr(arg, "bcs"):
-                    arg.bcs = []
-            arg.bcs.append(self)
-            return apply(self, *args, **kwargs)
+            annotate = annotate_tape(kwargs)
```

Comment:
This is absolutely essential for defcon. Before this PR, defcon reliably crashed whenever I tried to run long bifurcation analyses, killed by the out-of-memory manager. After this change, the same code runs happily for weeks. Thanks so much, @wence- !